libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartStoreExtensions.swift

```diff
@@ -0,0 +1,261 @@
+    case IndicesNotFoundError
+}
+
+extension SFSmartStore {
+    
+    public var Promises : SFSmartStorePromises {
+        return SFSmartStorePromises(api: self)
+    }
+    
+    public class SFSmartStorePromises {
+        
+        weak var api: SFSmartStore?
+        
+        init(api: SFSmartStore) {
+            self.api = api
+        }
+        
+        public func attributes(forSoup soupName: String) -> Promise<SFSoupSpec> {
+            return Promise(.pending) {  resolver in
+                let soupSpec : SFSoupSpec?  = self.api!.attributes(forSoup: soupName)
+                if let spec = soupSpec {
+                     resolver.fulfill(spec)
+                } else {
+                    resolver.reject(SmartStoreError.SoupNotFoundError)
+                }
+            }
+        }
+        
+        public func indices(forSoup soupName: String) -> Promise<[Any]> {
+            return Promise(.pending) {  resolver in
+                let indices : [Any]?  = self.api!.indices(forSoup: soupName)
+                if let indices = indices {
+                    resolver.fulfill(indices)
+                } else {
+                    resolver.reject(SmartStoreError.IndicesNotFoundError)
+                }
+            }
+        }
+        
+        public func soupExists(soupName: String) -> Promise<Bool> {
+            return Promise(.pending) {  resolver in
+                let result  = self.api!.soupExists(soupName)
+                resolver.fulfill(result)
+            }
+        }
+        
+        public func registerSoup(soupName: String, indexSpecs: [Any]) -> Promise<Bool> {
+            return Promise(.pending) {  resolver in
+                do {
+                   try self.api!.registerSoup(soupName, withIndexSpecs: indexSpecs, error: ())
+                   resolver.fulfill(true)
+                } catch let error {
+                    resolver.reject(error)
+                }
+            }
+        }
+        
+        public func registerSoup(soupSpec: SFSoupSpec,indexSpecs: [Any]) -> Promise<Bool> {
+            return Promise(.pending) {  resolver in
+                do {
+                    try self.api!.registerSoup(with: soupSpec, withIndexSpecs: indexSpecs)
+                    resolver.fulfill(true)
+                } catch  let error {
+                    resolver.reject(error)
+                }
+            }
+        }
+        
+        public func count(querySpec: SFQuerySpec) -> Promise<UInt> {
+            return Promise(.pending) {  resolver in
+                var count: UInt = 0
+                var error: NSError?
+                count = self.api!.count(with: querySpec, error: &error)
+                if let error = error {
+                    resolver.reject(error)
+                }else {
+                    resolver.fulfill(count)
+                }
+            }
+        }
+        
+        public func query(querySpec: SFQuerySpec, pageIndex: UInt) throws -> Promise<[Any]> {
```

Comment:
Should we add builders to make constructing SFSoupSpec, SFQuerySpec or SFSoupIndex easier?

Reply:
@wmathurin Yes I should have mentioned that it is pending. Do you see other objects that will benefit from the same pattern in the SmartStore?

Reply:
I can't think of any objects.